### PR TITLE
Add graph-safe ops and Inductor heuristics

### PR DESCRIPTION
**Pull request by froststeam**
## Summary

Add the graph-safe MUSA compatibility required by CUDA-facing inference paths and bump torchada from `0.1.83` to `0.1.85`:

- add MUSA `PrivateUse1` overrides for `aten::multinomial`, `aten::log`, and `aten::log_` on affected torch_musa releases
- reuse CUDA Inductor Triton template heuristics for MUSA matmul templates
- mirror `MUSA_VISIBLE_DEVICES` to `CUDA_VISIBLE_DEVICES` when both are present
- add focused correctness and CUDA-graph replay coverage for the new operators

## torch_musa version boundary

The TorchAda graph-safe compatibility is temporary and is limited to torch_musa versions before `2.11.0.post2`:

```text
torch_musa < 2.11.0.post2  -> keep TorchAda compatibility patches
torch_musa >= 2.11.0.post2 -> use torch_musa/ATen native implementations
```

The existing `_TORCH_MUSA_ACCELERATOR_FIX_VERSION = "2.11.0.post2"` is reused for this decision. The version is read from the existing `torch.musa.__version__` surface; no extra package-version helper or unconditional `import torch_musa` is added just for version detection.

On `torch_musa >= 2.11.0.post2`:

- `load_cpp_ops()` returns before building/loading TorchAda's C++ override extension
- the Inductor heuristic compatibility patch is skipped
- `torch_musa`/ATen owns `log`, `log_`, and `multinomial` dispatch directly
- no compile-time macro or version check is kept in `musa_ops.mu`

The currently published vLLM-MUSA image is still based on `torch_musa 2.11.0.post1+musa5.2.0`, so it remains on the compatibility side of this boundary.

## Scope

This PR does not change global CUDA platform semantics:

- does not patch `torch.cuda.is_available()`
- does not fake `torch.cuda.get_device_capability()`
- does not globally rewrite `torch.load` map locations
- does not add a flash-attention import shim
- does not add SGLang-Omni-specific model logic

## Testing

Focused validation was run with the PR source overlaid in the MUSA-enabled vLLM-MUSA image:

```bash
PYTHONPATH=src python -m pytest -q tests/test_cuda_patching.py \
  -k "VisibleDevicesEnv or InductorTemplateHeuristics or TestAcceleratorModuleWrapper or test_cpp_ops_loaded_on_musa"
```

```text
36 passed, 202 deselected
```

Operator correctness and graph replay tests on the same post1 MUSA runtime:

```bash
PYTHONPATH=src python -m pytest -q tests/test_log.py tests/test_multinomial.py
```

```text
8 passed
```

The version-boundary tests cover the existing post2 accelerator gate, post2 native Inductor registration behavior, local-version suffixes such as `+musa5.2.0`, and the post2+ no-C++-override behavior. `git diff --check` and Python syntax compilation also pass.

## Downstream image validation

| Downstream image | torch / torch_musa | Triton keys | Missing MUSA keys | Result |
| --- | --- | ---: | ---: | --- |
| `registry.mthreads.com/mcconline/inference/sglang:v0.5.12.post1-ph1-4.3.5-torch2.9.0-20260819` | `2.9.0` / `40305` | 11 | 0 | SGLang import and MUSA matmul passed after installing the image's optional runtime dependencies |
| `registry.mthreads.com/mcconline/inference/vllm:v0.24.0-ph1-5.2.0-torch2.11.0.post1-20260827` | `2.11.0.post1+musa5.2.0` / `50200` | 15 | 0 | vLLM import, MUSA platform plugin loading, and MUSA matmul passed |

A direct runtime probe in the latest image confirmed:

```text
torch_musa = 2.11.0.post1+musa5.2.0
torch.musa.is_available() = True
torch.musa.device_count() = 1
aten::log PrivateUse1 kernel = present
aten::log_ PrivateUse1 kernel = present
aten::multinomial PrivateUse1 kernel = present
```

## Request-level vLLM compile smoke

A request-level Qwen3-0.6B vLLM compile and GSM8K follow-up was run on MUSA with `CompilationMode.VLLM_COMPILE` (`mode=3`) using the compatible image `registry.mthreads.com/mcconline/inference/vllm:v0.24.0-ph1-5.2.0-torch2.9.1.post1-20260805`.

The service completed compilation, graph capture, a direct completion request, and the official 200-question GSM8K evaluator:

```text
Accuracy: 0.380
Invalid responses: 0.000
All 200 requests returned HTTP 200
```

The server log contained no runtime exception, backend compiler failure, graph break, or eager fallback during evaluation.

## Limitations

- No public `torch_musa 2.11.0.post2` wheel/tag is available yet; post2 behavior is covered by version-boundary tests and the code path is fail-safe for unknown versions.
- The `20260819` and `20260824` Qwen3 startup attempts hit a pre-existing vLLM-MUSA `is_neox` signature mismatch, so request-level evaluation used the compatible `20260805` image.
- The request-level run validates the Python Inductor path and downstream service behavior; it should not be interpreted as C++ override coverage for a post2 wheel.

### CODE REVIEW

File: src/torchada/_patch.py [hunk 1]
```diff
@@ -65,6 +65,17 @@ def _patch_something():
     return func
 
 
+@patch_function
+def _patch_visible_devices_env():
```
Comment (yeahdongcn) [inline src/torchada/_patch.py:69]: I think we also need to handle the case where both CUDA and MUSA are set. In that case, we should use the MUSA setting to override the CUDA one.

File: src/torchada/csrc/musa_ops.mu [hunk 2]
```diff
@@ -84,4 +356,11 @@ TORCH_LIBRARY_IMPL(aten, PrivateUse1, m) {
     // if (torchada::is_override_enabled("neg")) {
     //     m.impl("neg", torchada::neg_musa_impl);
     // }
```
Comment (yeahdongcn) [inline src/torchada/csrc/musa_ops.mu]: Again, this placeholder needs to be removed.

File: src/torchada/_patch.py [hunk 3]
```diff
@@ -108,6 +119,39 @@ def wrapper(*args, **kwargs):
     return decorator
 
 
+@patch_function
+@requires_import("torch._inductor.template_heuristics.registry")
+def _patch_inductor_template_heuristics():
```
Comment (yeahdongcn) [inline src/torchada/_patch.py:121]: This affects a lot of indirect behavior. Please add some unit tests for this and test it with both SGLang and vLLM-MUSA.

File: src/torchada/csrc/musa_ops.mu (hunk 2 above)
Comment (froststeam) [reply src/torchada/csrc/musa_ops.mu]: Done.

File: src/torchada/_patch.py (hunk 1 above)
Comment (froststeam) [reply src/torchada/_patch.py:69]: When both variables are set, MUSA_VISIBLE_DEVICES now takes precedence and overwrites CUDA_VISIBLE_DEVICES. I also updated the unit test to cover this behavior.

File: src/torchada/_patch.py (hunk 3 above)
Comment (froststeam) [reply src/torchada/_patch.py:121]: I added focused unit tests covering CUDA Triton heuristic copying, existing MUSA entries, cache invalidation, idempotency, malformed keys, and non-MUSA behavior (`7 passed`).

I also tested the PR with both downstream environments:

- SGLang / PyTorch 2.9: 11 CUDA Triton entries mapped to 11 MUSA entries, with no missing keys; SGLang import and MUSA matmul passed.
- vLLM-MUSA / PyTorch 2.11: 15/15 entries mapped, with no missing keys; vLLM import and MUSA plugin loading passed.
- Qwen3-0.6B with vLLM `torch.compile mode=3`: compilation and graph capture passed. A 200-question GSM8K run completed with all requests returning HTTP 200 and an invalid-response rate of 0.

The detailed commands, image versions, and results are included in the PR description.

File: src/torchada/csrc/musa_ops.mu [hunk 4]
```diff
@@ -67,6 +72,273 @@ at::Tensor neg_musa_impl(const at::Tensor& self) {
...
+            }
+            output[row * num_samples + sample] = chosen;
+        }
+        __syncthreads();
+    }
+}
+
+}  // namespace
+
+at::Tensor multinomial_musa_impl(
```
Comment (yeahdongcn) [inline src/torchada/csrc/musa_ops.mu:217]: This duplicates a full operator implementation at the MUSA kernel layer. ATen schemas/signatures, generator/RNG semantics, dtype handling, and edge-case behavior can change across torch versions, so this can keep compiling while silently diverging from upstream. Could we first reuse the torch_musa/ATen implementation and keep only a thin compatibility shim, or introduce an explicit versioned adapter? If a custom override is required, please add differential tests across the target torch versions (generator, replacement, invalid/NaN/negative inputs, empty and boundary shapes).

File: src/torchada/_patch.py [hunk 5]
```diff
@@ -108,6 +116,39 @@ def wrapper(*args, **kwargs):
...
```
Comment (yeahdongcn) [inline src/torchada/_patch.py:121]: This copies the current CUDA heuristic registry only once during import. That will miss lazy/future registrations, and it relies on private registry/cache names and key shape. Also, copying a CUDA heuristic class under a musa key does not establish that its lowering/template/autotune path is MUSA-compatible. Could we move this compatibility to the registration/lookup boundary (or use an explicit versioned allowlist) and add a real torch.compile + Inductor/Triton MUSA smoke after lazy imports? Unsupported templates/torch versions should fail closed or fall back rather than silently appearing supported.

Comment (lijing-mt) [reply src/torchada/_patch.py:121]: Current registration process is:
```
import torch._inductor.template_heuristics.registry as registry
----register all torch native "triton::" heuristics from `torch/_inductor/template_heuristics/triton.py`

......

from torch._inductor.codegen.common import init_backend_registration
init_backend_registration()
----lazy register all musa heuristics from `torch_musa/_inductor/template_heuristics.py`
```
When we need to fully import custom heuristics, it is recommended to modify `heuristic_registry` after calling `init_backend_registration()`.

Comment (froststeam) [reply src/torchada/_patch.py:121]: Implemented the registration-order change in commit `73be739`.

The patch now calls `torch._inductor.codegen.common.init_backend_registration()` before inspecting or modifying `_TEMPLATE_HEURISTIC_REGISTRY`. TorchMUSA's lazy native registrations therefore take precedence, and TorchAda only adds a CUDA-derived entry when the corresponding `(template, "musa", op_name)` key is still absent.

Validation in the newer vLLM-MUSA image:

- Image: `registry.mthreads.com/mcconline/inference/vllm:v0.24.0-ph1-5.2.0-torch2.11.0.post1-20260827`
- `torch_musa=2.11.0.post1+musa5.2.0`
- After `init_backend_registration()`: 7 native MUSA Triton heuristic keys were present; 8 CUDA keys remained without a MUSA counterpart.
- After TorchAda's fallback: all 15 corresponding MUSA keys were present, with native entries preserved.
- Focused tests, including a synthetic lazy-registration case: `38 passed`.

The existing `torch_musa >= 2.11.0.post2` gate still skips the TorchAda heuristic patch entirely, so fixed torch_musa versions use their native registration.

**Review by yeahdongcn — APPROVED** [review]
